gpbft/participant.go

```diff
@@ -157,57 +159,128 @@ func (p *Participant) beginInstance() error {
+		}
+		if err := power.Validate(); err != nil {
+			return nil, fmt.Errorf("invalid power table: %w", err)
+		}
+		c = &committee{power: power, beacon: beacon}
+		p.committees[instance] = c
+	}
+	return c, nil
 }
 
-func (p *Participant) handleDecision() error {
+func (p *Participant) handleDecision() {
 	if p.terminated() {
-		p.finalised = p.granite.terminationValue
-		p.terminatedDuringRound = p.granite.round
-		p.granite = nil
+		p.finalised = p.gpbft.terminationValue
+		p.terminatedDuringRound = p.gpbft.round
+		delete(p.committees, p.gpbft.instanceID)
+		p.gpbft = nil
+		p.currentInstance++
 		nextStart := p.host.ReceiveDecision(p.finalised)
 
 		// Set an alarm at which to fetch the next chain and begin a new instance.
 		p.host.SetAlarm(nextStart)
-		return nil
 	}
-	return nil
 }
 
 func (p *Participant) terminated() bool {
-	return p.granite != nil && p.granite.phase == TERMINATED_PHASE
+	return p.gpbft != nil && p.gpbft.phase == TERMINATED_PHASE
 }
 
 func (p *Participant) Describe() string {
-	if p.granite == nil {
+	if p.gpbft == nil {
 		return "nil"
 	}
-	return p.granite.Describe()
+	return p.gpbft.Describe()
+}
+
+// A power table and beacon value used as the committee inputs to an instance.
+type committee struct {
+	power  *PowerTable
+	beacon []byte
+}
+
+// A collection of messages queued for delivery for a future instance.
+// The queue drops equivocations and unjustified messages beyond some round number.
+type messageQueue struct {
+	maxRound uint64
+	// Maps instance -> sender -> messages.
+	// Note the relative order of messages is lost.
+	messages map[uint64]map[ActorID][]*GMessage
+}
+
+func newMessageQueue(maxRound uint64) *messageQueue {
+	return &messageQueue{
+		maxRound: maxRound,
+		messages: make(map[uint64]map[ActorID][]*GMessage),
+	}
+}
+
+func (q *messageQueue) Add(msg *GMessage) {
+	instanceQueue, ok := q.messages[msg.Vote.Instance]
+	if !ok {
+		// There's no check on instance number being within a reasonable range here.
+		// It's assumed that spam messages for far future instances won't get this far.
+		instanceQueue = make(map[ActorID][]*GMessage)
+		q.messages[msg.Vote.Instance] = instanceQueue
+	}
+	// Drop unjustified messages beyond some round limit.
+	if msg.Vote.Round > q.maxRound && isSpammable(msg) {
+		return
+	}
+	// Drop equivocations and duplicates (messages with the same sender, round and step).
+	for _, m := range instanceQueue[msg.Sender] {
+		if m.Vote.Round == msg.Vote.Round && m.Vote.Step == msg.Vote.Step {
+			return
+		}
+	}
+	// Queue remaining good messages.
+	instanceQueue[msg.Sender] = append(instanceQueue[msg.Sender], msg)
+}
+
+func (q *messageQueue) Drain(instance uint64) []*GMessage {
+	var msgs []*GMessage
+	for _, ms := range q.messages[instance] {
+		msgs = append(msgs, ms...)
+	}
+	// Sort by step so messages will be processed in a useful order.
+	sort.Slice(msgs, func(i, j int) bool {
```

Comment:
Do we care about the sort stability? We probably should for testing purposes.

Reply:
No harm in using a stable sort 👍 . But also this should sort by round too!